File: crates/topcoat-view/src/runtime/reactive/expr/expr_lit.rs
use crate::runtime::{Expr, Interpreter, IntoExpr};

pub struct ExprLit<T>(T);

impl<T> Expr for ExprLit<T> {
    type Output = T;

    fn evaluate(self, _interpreter: &mut Interpreter) -> Self::Output {
        self.0
    }
}

macro_rules! impl_primitive {
    ($ty:ty) => {
        impl IntoExpr for $ty {
            type Expr = ExprLit<Self>;

            fn into_expr(self) -> Self::Expr {
                ExprLit(self)
            }
        }
    };
}

impl_primitive!(bool);
impl_primitive!(f64);
impl_primitive!(&'static str);

File: crates/topcoat-runtime/browser/src/expr/expr_lit.ts
import type { Interpreter } from "./interpreter";

export type ExprLit<T> = {
	type: "Lit";
	value: T;
};

export function eval_expr_lit<T>(
	expr: ExprLit<T>,
	_interpreter: Interpreter,
): T {
	return expr.value;
}
